packages/core/src/scuttle.js

```diff
@@ -0,0 +1,181 @@
+/**
+ * @typedef {object} ScuttleOpts
+ * @property {boolean} enabled - Whether scuttling is enabled or not.
+ * @property {Array<string|RegExp>} exceptions - List of properties to exclude from scuttling.
+ * @property {string} scuttlerName - Name of the scuttler function to use.
+ * @property {(globalRef: object, scuttle: Function) => void} scuttlerFunc - Scuttler function to use.
+ */
+
+/**
+ * @typedef {object} GlobalRef
+ * @property {object} [globalThis] - Reference to the global object.
+ */
+
+const { Object, Array, Error, RegExp, Set, console, Proxy, Reflect } =
+  globalThis
+
+const {
+  assign,
+  getOwnPropertyNames,
+  getOwnPropertyDescriptor,
+  create,
+  defineProperty,
+} = Object
+
+const { isArray, from } = Array
+
+const { getPrototypeOf } = Reflect
+
+const { warn } = console
+
+/**
+ * Applies scuttling, with the default set of options, including using Snow if passed in as scuttlerFunc.
+ * Scuttle globalThis right after we used it to create the root package compartment.
+ *
+ * @param {GlobalRef} globalRef - Reference to the global object.
+ * @param {ScuttleOpts} opts - Scuttling options.
+ */
+function scuttle(globalRef, opts) {
+  const scuttleOpts = generateScuttleOpts(globalRef, opts)
+
+  if (scuttleOpts.enabled) {
+    if (!isArray(scuttleOpts.exceptions)) {
+      throw new Error(
+        `LavaMoat - scuttleGlobalThis.exceptions must be an array, got "${typeof scuttleOpts.exceptions}"`
+      )
+    }
+    scuttleOpts.scuttlerFunc(globalRef, (realm) =>
+      performScuttleGlobalThis(realm, scuttleOpts.exceptions)
+    )
+  }
+}
+
+/**
+ * @param {GlobalRef} globalRef - Reference to the global object.
+ * @param {ScuttleOpts|boolean} originalOpts - Scuttling options. Accepts `true` for backwards compatibility.
+ * @returns {ScuttleOpts} - Final scuttling options.
+ */
+function generateScuttleOpts(globalRef, originalOpts = create(null)) {
+  const defaultOpts = {
+    enabled: true,
+    exceptions: [],
+    scuttlerName: '',
+  }
+  const opts = assign(
+    {},
+    originalOpts === true ? { ...defaultOpts } : { ...originalOpts },
+    { scuttlerFunc: (globalRef, scuttle) => scuttle(globalRef) },
+    {
+      exceptions: (originalOpts.exceptions || defaultOpts.exceptions).map((e) =>
+        toRE(e)
+      ),
+    }
+  )
+  if (opts.scuttlerName) {
+    if (!globalRef[opts.scuttlerName]) {
+      throw new Error(
+        `LavaMoat - 'scuttlerName' function "${opts.scuttlerName}" expected on globalRef.` +
+          'To learn more visit https://github.com/LavaMoat/LavaMoat/pull/462.'
+      )
+    }
+    opts.scuttlerFunc = globalRef[opts.scuttlerName]
+  }
+  return opts
+
+  /**
+   * @param {string|RegExp} except - Exception to convert to RegExp.
+   * @returns {string|RegExp} - Converted exception.
+   */
+  function toRE(except) {
```

Comment:
consider using [escape-string-regexp](https://www.npmjs.com/package/escape-string-regexp)?

Reply:
I don't think it provides the service I'm looking for here..
Also I think it is preferred we don't rely on third party services for production code unless we really have to.
Apparently that isn't the case entirely in LavaMoat atm, but still

Reply:
If you're converting an arbitrary string to a regexp, you will need to escape it properly.  If there's a discrete list of strings we expect, we could keep a `Map<string, RegExp>` or something. if we don't know what those strings are, then we have more work to do.

I didn't realize how simple that package was.  but it seems like we could adapt the regex for our special case here (it looks like there's some sort of convention in use??)

```js
	return string
		.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&')
		.replace(/-/g, '\\x2d');
```

Reply:
Can you give me an example of why might this be needed?
While we don't know the strings we might receive, we know they're not going to include any special chars other than `abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_` because they represent names of props we expect to find on the global object.
Is it possible potentially to find a "🐴" in a property name of the global object? Yes.
Will it ever happen though?

Reply:
Either way, fixed aba162b32

Reply:
Also, [it](https://github.com/sindresorhus/escape-string-regexp/blob/main/index.js) doesn't seem to do what I would expect it to:
![Screenshot 2024-01-29 at 14 37 01](https://github.com/LavaMoat/LavaMoat/assets/13243797/2e4d6ad7-c6c2-4203-91e3-e670f2114e74)

Reply:
I didn't test the thing I pasted. 🙂	

A property name can be literally any string in JS, afaik.  Some of those strings contain characters that have special meaning to regular expressions.  That's my concern, but as long as it's not exposing some sort of ReDoS-style flaw, then I'm not sure we'd need to worry too much about it (until we do).